Added esupgrade pre-commit hook

codingjoe opened the pull request:
> https://fosstodon.org/@codingjoe/115803674715881666
> @pauloxnet like this?

* Add esupgrade to pre-commit hooks
* pre-commit CI applies safe changes
* small unrelated changes, co-pilot discovered
* set explicit console log level (debug)

codingjoe commented on djangoproject/static/js/fundraising-heart.js:
```diff
@@ -45,48 +45,50 @@
-      var percent = this.heart.dataset.percent;
-      var fadedCount = Math.ceil((this.pixels.length * (100 - percent)) / 100);
-      for (var i = 0; i < fadedCount; i++) {
+      const pixels;
```
pixels is reassigned later.
```suggestion
      let pixels;
```

codingjoe commented on djangoproject/static/js/dashboard/detail.js:
```diff
@@ -1,24 +1,24 @@
+  $.getJSON(url, (response) => {
+    for (const i = 0; i < response.data.length; i++) {
       response.data[i][0] = response.data[i][0] * 1000;
     }
```
This should be transformed into a for of loop.

codingjoe commented:
Ok, there is some name shadowing happening in one file that threw me off. I reviewed all changes again, and they all appear correct.

If anyone thinks this is helpful, go ahead. Otherwise, just close the PR. I am happy to have tested such a history-rich code base. 😉

pauloxnet commented:
> @pauloxnet like this?

Thanks for the PR

I'll ask @adamzap and other @django/django-website to review it and share their feedback

codingjoe commented:
OK, let me address the co-pilot changes too in a separate commit.

ulgens commented:
> I can split the manual changes into a separate PR if that makes reviewing them easier.

I don't think they make an important difference for the review; it would be a small improvement for the git history. (It was a nitpick 🌻 )

adamzap commented on djangoproject/static/js/dashboard/index.js:
```diff
@@ -1,13 +1,13 @@
 
-    var url = element.data('path') + element.data('metric') + '.json';
-    $.getJSON(url, function (response) {
+    const url = `${element.data('path') + element.data('metric')}.json`;
```
Same here as above.

adamzap commented on djangoproject/static/js/mod/list-collapsing.js:
```diff
@@ -30,30 +28,31 @@ define([
-        var pos = $(hash).position();
-        $(window).scrollTop(pos.top);
+        const pos = $(hash).position();
+        $(globalThis).scrollTop(pos.top);
```
Is using `globalThis` a best practice for web-only code? We don't need to support any other JavaScript environments.

codingjoe commented on djangoproject/static/js/dashboard/detail.js:
```diff
@@ -1,24 +1,24 @@
-    show: function (x, y, message) {
+$(() => {
+  const element = $('#graph');
+  const url = `${element.data('path') + element.data('metric')}.json?days=365`;
```
Agreed, this transformation isn't automated by esupgrade, since we don't know that element.data doesn't return an integer (where addition would work differently).

codingjoe commented on .pre-commit-config.yaml:
```diff
@@ -40,6 +40,12 @@ repos:
           - id: django-upgrade
             args: [--target-version, "6.0"]
 
+    - repo: https://github.com/codingjoe/esupgrade
```
None taken; this was inspired by a conversation on Mastodon and meant as a testing ground to harden esupgrade.

I can understand hesitance to introduce this tool to the dev toolchain, and I don't mind either way.

However, I reviewed all automatic changes and, of course, my own in the latter commits.
So if those changes find appreciation, I am happy to make this a one-time thing.

codingjoe commented on djangoproject/static/js/djangoproject.js:
```diff
@@ -1,45 +1,41 @@
-// Toggle persistent display of documentation version and language options
```
I rebase with the latest version and the latest changes, which are now including the comments.

codingjoe commented on djangoproject/static/js/dashboard/detail.js:
```diff
@@ -41,27 +41,22 @@ $(function () {
-        unit
-      );
+    const format_message = (timestamp, measurement) => {
+      const unit = measurement == 1 ? response.unit : response.unit_plural;
```
There are numerous places with incorrect equality checks. This is a case for ESLint.

codingjoe commented:
@ulgens @adamzap thanks for your reviews! I updated everything and split the pre-commit config change into a separate commit. So you could go ahead and just cherry-pick the latter commits.

There are numerous blunders in the JS code, none of which are caused by this transformation. I set up ESLint locally… and oh boy… I'd love to fix it, but in a separate PR. Let's get things out of the door in baby steps.

codingjoe commented:
Sooooo, where are with this? Just given the extra commit added after my rebase might hint towards this being somewhat useful :)

codingjoe commented:
For whatever reason, I opened this PR on my main branch. To improve my workflow, I will close and reopen the PR with a link to this discussion. Sorry for the inconvenience.
